sdk/program/src/vote/state/mod.rs

```diff
@@ -374,50 +374,58 @@ impl VoteState {
         3762 // see test_vote_state_size_of.
     }
 
-    // we retain bincode deserialize for not(target_os = "solana")
-    // because the hand-written parser does not support V0_23_5
+    /// Deserializes the input buffer into a newly allocated `VoteState`
+    ///
+    /// This function is intended as a drop-in replacement for `bincode::deserialize()`.
+    /// V0_23_5 is not supported in a BPF context, but all versions are supported on non-BPF.
     pub fn deserialize(input: &[u8]) -> Result<Self, InstructionError> {
-        #[cfg(not(target_os = "solana"))]
-        {
-            deserialize::<VoteStateVersions>(input)
-                .map(|versioned| versioned.convert_to_current())
-                .map_err(|_| InstructionError::InvalidAccountData)
-        }
-        #[cfg(target_os = "solana")]
-        {
-            let mut vote_state = Self::default();
-            Self::deserialize_into(input, &mut vote_state)?;
-            Ok(vote_state)
-        }
+        let mut vote_state = Self::default();
+        Self::deserialize_into(input, &mut vote_state)?;
+        Ok(vote_state)
+    }
+
+    // this only exists for the sake of the feature gated upgrade to the new parser; do not use it
+    #[doc(hidden)]
+    pub fn deserialize_with_bincode(input: &[u8]) -> Result<Self, InstructionError> {
+        deserialize::<VoteStateVersions>(input)
+            .map(|versioned| versioned.convert_to_current())
+            .map_err(|_| InstructionError::InvalidAccountData)
     }
```

Comment:
i waffled a lot on whether to try and do the feature gate and parser upgrade all in this pr or split into two, but the fact that some functions that want to use `VoteState::deserialize()` dont have access to the featureset struct made me decide the changes are too complicated and should be done separately